### ISSUE TITLE
User Experience Design - initial documents

### ISSUE THREAD

**Issue by kytrinyx** [exercism/discussions#154]
We have published five new documents that make up the initial exploration of the foundational, philosophical questions that Exercism is built on.

The starting point for the discussion is motivation, and the elements that strengthen or weaken motivation. It explores how we can tap into people's goals and values, communicating Exercism's vision effectively to better serve those whose goals overlap with ours, the key points in people's emotional journey where negative emotions might cause them to stop using Exercism, and how progression, completion, and success play in.

[Please take some time to read through the documents, and reflect on how these match your experiences with Exercism.](https://github.com/exercism/docs/blob/master/about/conception/README.md)

**Requested Feedback & Thoughts**

We are aware that there will be quite large implications in terms of how Exercism language tracks will be structured, and also for how we support code review on the site. The documents do not delve into the practical outcomes of these fundamental choices and principals. In other words: this discussion is not about implementation or features.

There are a number of important questions that are not covered here, some of which we have asked but not yet explored, some which we are still trying to properly formulate, and others that we are discussing amongst the reboot team.

What we'd most like from you with respect to these documents is reactions to and thoughts about the contents within them, rather than ideas about how these things might be implemented in practice.

In particular:

* Are we missing any key insights?
* Are there questions that you believe tie directly to these topics, which we should consider before moving forward?
* Do you have any concerns with our conclusions?

Thank you 🌻

**Comment by willnwhite** [comment]
In https://github.com/exercism/docs/blob/master/about/conception/progression.md:

> Think back to school when you were were doing maths homework and had to wade through hundreds of examples in algebra. While many of us probably took some perverse delight in this, for most students this sort of practice is painful and disheartening because it is hard to gauge progress.

My maths textbooks had the answers in the back, so it wasn't hard to gauge progress. Don't they all? This example just doesn't ring true for me. 

I'm not sure if this informs the debate or not.

**Comment by iHiD** [comment]
> My maths textbooks had the answers in the back, so it wasn't hard to gauge progress. Don't they all? This example just doesn't ring true for me.

@willnwhite To clarify what I meant when I wrote this (rather than to agree or disagree with your comment)...

I guess in my head there is a difference between progress in terms of:
- Completing and getting things right (e.g. doing exercise and looking at answers at the back)
- Knowing whether you're actually understanding more and getting better.

Specifically the next sentence clarifies what I meant by that I think: "**Variety** is in short supply and **boredom** easily sets in." When I just do the same things lots of times and the only measure of progress is right/wrong or getting through a list of questions, that doesn't feel heartening for me. I need to understand the bigger picture of **how** these equations will help me progress towards something **bigger** than the equations themselves. Without a structured timeline that things fit into, practice can very quickly feel like practice for the sake of practice.

**Comment by emilianbold** [comment]
I disagree with the Code Review conclusion:

> Based on the above points, we believe that the right choice for Exercism to restrict the giving of feedback to mentors. However, we feel that it is essential that the benefits of community feedback are not lost. Specifically we believe it is crucial to provide an onramp so that anyone who wants to can train to become a mentor. Mentorship must not be elitist and must not be restricted to the "superstars" of communities. Throughout the site, the option to become a mentor must be clear and welcoming.

I believe you underlined most of my reasoning in the 'Pros of Community Feedback' section.

Having mentors is useful, but I want to be able to give and receive feedback.

One interesting thing I don't believe you noticed is that people learn in batches or waves. So, at least initially, you have almost the same set of people submitting solutions. They are like a class or a study group. Why shouldn't they review each other's solutions?

Sure, you might have a bad study group but the chances are slim. And you can detect if there is a single user posting negative comments all over the place.

These are my 2c.

Honestly, I just wanted some exercises. The scope of what you are trying to accomplish here seems very large.

**Comment by kytrinyx** [comment]
Thank you @willnwhite. I want to make sure that I understand your perspective correctly!

> Having mentors is useful, but I want to be able to give and receive feedback

From our perspective anyone can sign up to become a mentor. Mentorship isn't a 1:1 thing, we want to have a community of folks who give feedback.

I suppose my question is: if anyone can sign up to participate in giving feedback, does that still feel like you can't give feedback?

Do you currently give feedback?

> They are like a class or a study group. Why shouldn't they review each other's solutions?

Study groups are a different thing, I think. We haven't dug into this, but we currently have teams, and I expect that we would continue to have that (not sure about the details about how it would work, but I imagine that within a team it functions like a study group).

> And you can detect if there is a single user posting negative comments all over the place.

I would like to hear more about how you would suggest doing this.

**Comment by emilianbold** [comment]
> From our perspective anyone can sign up to become a mentor.[...]
> I suppose my question is: if anyone can sign up to participate in giving feedback, does that still feel like you can't give feedback?

Signing up to become a mentor seems like a big commitment, following some training track, etc. etc.

If anyone can give feedback by clicking Yes to some dialog, it seems a bit silly to even have that dialog.

> Do you currently give feedback?

Yes, after I post a solution I try to read the latest 5-10 solutions and I comment on or 'like' a few of them. Reading alternative solutions seems to me just as important as doing my own.

Just last night I spent more time actually trying to understand somebody else's solution than I spent doing my own. Blew my mind the language allowed you to do that.

> Study groups are a different thing, I think. We haven't dug into this, but we currently have teams [...]

I have gotten to know (by avatar) at least 4-5 other people that seem to post solutions and learn the same language as me. We are not a formal team, but I see their solutions next to mine, read them and comment.

>> And you can detect if there is a single user posting negative comments all over the place.

> I would like to hear more about how you would suggest doing this.

Not sure what you are focusing here. The 'negative' part might be a 'Flag this post' kind of thing, maybe with some meta-moderation (from admins / mentors).

But I was actually thinking of locality. Somebody starting to learn a language in May will not dig until last November for a solution to comment on. You have the data, you can plot some graphs and see if it checks out or not. Do some (machine learning) models, have bots check for warning signs, etc.

You could also auto-archive a solution after, say, 1 month, particularly if the user is already 10 steps ahead -- they probably don't care for feedback going back that far.

**Comment by kytrinyx** [comment]
> Signing up to become a mentor seems like a big commitment, following some training track, etc. etc.

Our experience with the feedback is that there are HUGE differences in the quality of the feedback. I think for many people the training might not require a lot from them, but would make a huge difference. For other people, learning the skills might be a lot more challenging. Either way, based on the last four years, I am very wary of continuing in the path that we are currently on.

> If anyone can give feedback by clicking Yes to some dialog, it seems a bit silly to even have that dialog.

It's not going to be a simple dialog—if it were the case, I absolutely agree. Then we would simply go with the anyone can give feedback.

> Reading alternative solutions seems to me just as important as doing my own.

Agreed, that's how I learn. I've rarely received feedback, and learn a tremendous amount from looking at other people's code.

> I have gotten to know (by avatar) at least 4-5 other people that seem to post solutions and learn the same language as me. We are not a formal team, but I see their solutions next to mine, read them and comment.

This is really interesting, and I'm going to take that back to the design team, as I think there's a kernel of something important going on there.

> Do some (machine learning) models, have bots check for warning signs, etc.

The head of the team who is doing the user experience design with us has a degree in AI, and is doing spectacularly interesting work in other fields with it. We definitely plan on using various approaches to finding better ways to show people interesting data, detect trends, etc.

Thanks for sharing your thoughts, this has given me a lot to think about and a few things to take back to the design team. 🌷

**Comment by NobbZ** [comment]
Only skimming this discussion, but it reads as if I'm not even able to ask
questions about a piece of code I've seen from another user unless I'm a
mentor? Beeing able to ask was always the main reason for me to look at
other people's code, not to show them what they've done wrong...

Katrina Owen <notifications@github.com> schrieb am Di., 6. Juni 2017, 04:47:

**Comment by iHiD** [comment]
> Yes, after I post a solution I try to read the latest 5-10 solutions and I comment on or 'like' a few of them. Reading alternative solutions seems to me just as important as doing my own. 

**tl;dr; IMO, The discovery and discussion of solutions is crucial and should be open to all. However, the discussion during someone completing an exercise should be restricted to people who have shown skill at giving feedback in a positive and productive way (ie "mentors").**

This is really interesting. So I think there are two very distinct concepts here that we're compounding:
1) Giving feedback *during* the process of someone completing an exercise, when they're in the learning phase of that exercise.
2) Commenting and discussion someone's solution *after* they've completed it.

In my head, having trained, trusted mentors giving feedback during the learning phase is essential. All our research has shown that ~50% of people dislike code-review and feedback because they feel it's done in a bad way. The only way to fix that is to vet/train mentors first. The on-ramp is crucial, so that anyone can become a mentor, but having a barrier of quality is crucial.

The second, however, is a very different situation. Being able to publish a solution and then engage in discussion about it, once a mentor has helped you through it, is a really valuable phase, and seems to be the area that the comments above are concerned about losing. We've not addressed that in these documents at all, and so I think they read like we're not going to allow discussion amongst the community. That's not our intention. The favouring and discussing of solutions is valuable and something I think we should continue to incorporate into the flow of Exercism.

@emilianbold @NobbZ I hope that adds some clarity. Thank you for helping raise that distinction -  we hadn't made clear well enough in the documents.

**Comment by NobbZ** [comment]
But after a mentor has guided the student to a "better" solution, the
aspect I had asked about might have been refactored away. Having only
polished and uniformized solutions open to the public might remove most of
the things I loved in exercism.

I can see though that the current way might scare new students away,
because of misunderstandings. But those misunderstandings will still happen
when we have proper taught mentors. Perhaps just because english isn't the
first language of one of the participants of the discussion. Perhaps
students just leave because they thaught of exercism as a codewars or
similar, where you are ranked for the performance of your code? Perhaps
they just leave because they don't like that tone of pink?

Most leaving people don't tell why they leave, but people that stay might
tell you why they are still there, and they even might tell you that they'd
leave if certain changes are introduced.

And to be honest, I'm near to leaving if we have mentors only commenting,
but I'd try how it feels for a month or so.

I really like to get feedback, I'm really happy when I get notified about
likes, but even more I am happy when I ask on another solution to explain
feature X of a language and a discussion about pro and cons evolves.

I like the diversity of approaches through a track, and I really fear this
will get lost under the "dictatorship" of a handfull of track mentors.



Jeremy Walker <notifications@github.com> schrieb am Di., 6. Juni 2017 um
14:19 Uhr:

**Comment by iHiD** [comment]
> I really like to get feedback, I'm really happy when I get notified about likes, but even more I am happy when I ask on another solution to explain feature X of a language and a discussion about pro and cons evolves.

That feels like its at the core of what you're saying, and I think it's a valuable thing that should not be lost. There's a subtlety here though. 

1) If you're asking someone else to explain feature X of a language, you're considering that they know more about it than you. In that situation, then commenting and asking after the exercise has been completed is fine. You're effectively the learner, chatting and discussing with someone who has different knowledge. We want to encourage and support that behaviour.
2) In the other situation, though, where someone is a beginner and using a "too basic" feature of a language, what's important is that they receive good feedback and support to help them learn a different feature without feeling stupid or ignorant, and without becoming discouraged.

The core value of Exercism is having a safe place for people to learn. We therefore feel that protecting the quality of that feedback is essential. However, there is no reason that you couldn't look at early "non-refactored" versions of someone's solution and have an interesting discussion on it. That's a really valuable thing.

> I like the diversity of approaches through a track, and I really fear this will get lost under the "dictatorship" of a handful of track mentors.

I agree with this too! It's essential that we don't have a dictatorship of track mentors. However, I'm not sure why we'd end up with that. The core values of mentorship on Exercism will be about helping and teaching in a compassionate way - people with dictatory tendencies are the exact people we *don't* want becoming mentors and that we want to filter out through the onboarding process.

**Comment by kytrinyx** [comment]
I think one of the things that is not yet clearly articulated (because I'm still trying to find the shape of the idea in my head) is that the good feedback that we are looking for is the opposite of dictatorial.

Some of the problematic feedback that we see is in a style "X is wrong" "Do Y". Sometimes (occasionally) something is actually "wrong" in the sense that the language has very strong preferences towards something else, or because it causes a race condition... but the unfortunate feedback doesn't help the learner discover this, it simply dictates a different approach or a solution.

More useful feedback will take trade-offs into account, and allow for a much broader range of approaches, and will center the discussion around _why_ one approach might be more desirable than another.

I would expect that vetting people and training people in the skills of code review would cause solutions to be more diverse, not less so.

Of course, that can not be verified until we have actually tried it.

I think you are right, @NobbZ, that having the opportunity to study solutions and discuss them is vitally important, and I hope that we are able to preserve that and also that we are able to improve the quality of the feedback for everyone.

**Comment by ErikSchierboom** [comment]
Initially, I also felt that only allowing a fixed set of users might not be a great idea. However, I then started to think on it and I started to like it more and more. I actually think it's a great idea now! Let me explain why I think it is a good idea.

- Having a set of mentors will lead to dedicated users that are invested in commenting, meaning frequent comments. This will almost certainly lead to more comments, which to me is *the* selling point of the mentors approach. I have completed several tracks, and there was a huge difference in number of comments between the tracks. Those with more comments were definitely much more helpful to me.
- The mentors will have been trained to give good feedback, which sounds easy but absolutely isn't! I think most people will have noticed that it is terribly easy to have a textual (online) comment (e.g. in a mail) be misinterpreted by the other person. As the mentors will have had specific training for this, the chance of this happening will be much lower. Also, correctly worded feedback will allow users to better focus on the contents of the comment, not on its form.
- The mentors will have a good grasp of the language involved. This will lead to high quality comments that are likely to really help people.
- The mentors will also be invested to comment on *less interesting* issues and keep on doing that, which other users might not bother with (or stop bothering with after a while).
- The mentors will have a good understand of the complete track, and can thus refer to relevant exercises and such, of which other users cannot or might not be aware.

Most of the above points are not necessarily tied to being a mentor, the "average" user might also have them. However, the fact that the mentors will have been trained to give feedback and must have a good grasp of the language, makes it more likely for them to have these qualities.

It is also a bit of a fundamental choice: do you want to use Exercism to discuss your and other people's code or do you want to use Exercism to improve your fluency/proficiency in a language? In the first case, the mentorship idea doesn't really make sense, and you would probably want Exercism to be like a forum where everyone can post. However, in the second case, having a set of mentors most assuredly does make sense. Think about it. If you are learning a language, would you rather get comments from a smaller set of people with deep knowledge of the language that are trained to give feedback (like good school teachers), or from a larger set of people with possibly less knowledge of the language that are not trained to give feedback (like students). I really do think the former would be much better.

Of course, people shouldn't have the feeling that it's an ivory tower, so some way of communicating with mentors could be useful.

The people involved in rethinking the user experience will have known that the mentors setup would meet with a lot of resistance, and I was initially skeptical. However, having giving it some thought, I am actually very excited about the new setup. I think it will lead to more and more high quality comments, which will greatly help people in improving fluency in their language of choice.

**Comment by emilianbold** [comment]
> It is also a bit of a fundamental choice: do you want to use Exercism to discuss your and other people's code or do you want to use Exercism to improve your fluency/proficiency in a language?

It's not a choice the way you present it. Some people might prefer to "use Exercism to improve your fluency/proficiency in a language" **BY** "discuss[ing] your and other people's code".

Anyhow, once mentors are in, I guess people that prefer mentors, like @ErikSchierboom, will only talk to mentors. Which I don't mind.

Only problem is if there is no other way except talking to mentors.

I guess it could even be a matter of permissions "Allow comments from mentors only", "Allow comments from my team", "Allow comments from my study group", "Allow comment from everybody". But instead of tweaking settings it would be better to have a single Exercism 'experience' which does allow everybody to comment, like today.

Of course, just my 2c, etc.

**Comment by ErikSchierboom** [comment]
> It's not a choice the way you present it. Some people might prefer to "use Exercism to improve your fluency/proficiency in a language" BY "discuss[ing] your and other people's code".

Ah yes, but that's how Exercism currently works! And arguably it doesn't actually work as well as you might hope. But I see your point, although I am still in favor of mentors.

**Comment by iHiD** [comment]
@emilianbold I appreciate all your comments. It's driven lots of thoughts and discussion amongst the team working on this today and made us consider things from angles we hadn't looked at in as much depth before.

> Only problem is if there is no other way except talking to mentors.

I suspect that you'll find that your experience changes very little. The only difference you should notice that the feedback you receive while you're doing an exercise comes from mentors. You'll still be able to comment on other people's submissions (if they allow that), just after they have completed the exercise, not during their learning process. So if your goal is to *"improve your fluency/proficiency in a language by discussing your and other people's code* I still think you'll be entirely able to do that. 

We want to encourage the exercism community to learn together, not restrict that. But we also want to ensure that the conversation is not offputting, something that is a common problem in both online communities in general, and also with code-review in general.

**Comment by kytrinyx** [comment]
Alrighty, the new prototype is up and getting feedback. I'm going to go ahead and close this.

https://github.com/exercism/v2-feedback/blob/master/README.md

❤️ thanks everyone for all your feedback and thoughts along the way. This has been incredibly helpful.
